File: internal/models/access_admin.go
package models

import (
	"encoding/json"
	"time"
)

type PermissionGrant struct {
	ID          string `json:"id,omitempty"`
	Resource    string `json:"resource"`
	Action      string `json:"action"`
	Description string `json:"description,omitempty"`
}

type ManagedRole struct {
	ID             string            `json:"id"`
	OrganizationID *string           `json:"organization_id,omitempty"`
	Name           string            `json:"name"`
	Slug           string            `json:"slug"`
	Description    string            `json:"description,omitempty"`
	IsSystemRole   bool              `json:"is_system_role"`
	IsCustom       bool              `json:"is_custom"`
	Version        int64             `json:"version"`
	CreatedBy      *string           `json:"created_by,omitempty"`
	UpdatedBy      *string           `json:"updated_by,omitempty"`
	Permissions    []PermissionGrant `json:"permissions"`
	AssignedUsers  int               `json:"assigned_users"`
	CreatedAt      time.Time         `json:"created_at"`
	UpdatedAt      time.Time         `json:"updated_at"`
	DeletedAt      *time.Time        `json:"deleted_at,omitempty"`
}

type ManagedRoleCreateInput struct {
	Name        string            `json:"name"`
	Slug        string            `json:"slug,omitempty"`
	Description string            `json:"description,omitempty"`
	Permissions []PermissionGrant `json:"permissions"`
}

type ManagedRolePatch struct {
	Name             *string            `json:"name,omitempty"`
	Slug             *string            `json:"slug,omitempty"`
	Description      *string            `json:"description,omitempty"`
	ClearDescription bool               `json:"clear_description,omitempty"`
	Permissions      *[]PermissionGrant `json:"permissions,omitempty"`
	ExpectedVersion  int64              `json:"expected_version"`
}

type ManagedRoleCloneInput struct {
	Name        string `json:"name"`
	Slug        string `json:"slug,omitempty"`
	Description string `json:"description,omitempty"`
}

type ManagedRoleListFilter struct {
	PaginationRequest
	Search        string
	IncludeSystem bool
}

type ManagedRoleImpact struct {
	RoleID              string            `json:"role_id"`
	AssignedUsers       int               `json:"assigned_users"`
	CurrentPermissions  int               `json:"current_permissions"`
	ProposedPermissions int               `json:"proposed_permissions"`
	Added               []PermissionGrant `json:"added"`
	Removed             []PermissionGrant `json:"removed"`
}

type ManagedRoleAssignment struct {
	RoleID     string    `json:"role_id"`
	UserID     string    `json:"user_id"`
	Email      string    `json:"email"`
	FirstName  string    `json:"first_name"`
	LastName   string    `json:"last_name"`
	AssignedBy *string   `json:"assigned_by,omitempty"`
	AssignedAt time.Time `json:"assigned_at"`
}

type ManagedRoleAssignmentInput struct {
	UserID string `json:"user_id"`
	Reason string `json:"reason"`
}

type ManagedRoleUnassignmentInput struct {
	Reason string `json:"reason"`
}

type RoleChangeEvent struct {
	ID           string          `json:"id"`
	RoleID       string          `json:"role_id"`
	TargetUserID *string         `json:"target_user_id,omitempty"`
	EventType    string          `json:"event_type"`
	ActorUserID  string          `json:"actor_user_id"`
	RoleVersion  int64           `json:"role_version"`
	Reason       string          `json:"reason,omitempty"`
	BeforeState  json.RawMessage `json:"before_state,omitempty"`
	AfterState   json.RawMessage `json:"after_state,omitempty"`
	CreatedAt    time.Time       `json:"created_at"`
}

File: frontend/src/types/access-admin.ts
export interface PermissionGrant {
  id?: string;
  resource: string;
  action: string;
  description?: string;
}

export interface ManagedRole {
  id: string;
  organization_id?: string;
  name: string;
  slug: string;
  description?: string;
  is_system_role: boolean;
  is_custom: boolean;
  version: number;
  created_by?: string;
  updated_by?: string;
  permissions: PermissionGrant[];
  assigned_users: number;
  created_at: string;
  updated_at: string;
  deleted_at?: string;
}

export interface ManagedRoleCreateInput {
  name: string;
  slug?: string;
  description?: string;
  permissions: PermissionGrant[];
}

export interface ManagedRolePatch {
  name?: string;
  slug?: string;
  description?: string;
  clear_description?: boolean;
  permissions?: PermissionGrant[];
  expected_version: number;
}

export interface ManagedRoleCloneInput {
  name: string;
  slug?: string;
  description?: string;
}

export interface ManagedRoleListParams {
  search?: string;
  include_system?: boolean;
  page?: number;
  page_size?: number;
}

export interface ManagedRoleImpact {
  role_id: string;
  assigned_users: number;
  current_permissions: number;
  proposed_permissions: number;
  added: PermissionGrant[];
  removed: PermissionGrant[];
}

export interface ManagedRoleAssignment {
  role_id: string;
  user_id: string;
  email: string;
  first_name: string;
  last_name: string;
  assigned_by?: string;
  assigned_at: string;
}

export interface ManagedRoleAssignmentInput {
  user_id: string;
  reason: string;
}

export interface ManagedRoleUnassignmentInput {
  reason: string;
}

export interface RoleChangeEvent {
  id: string;
  role_id: string;
  target_user_id?: string;
  event_type: string;
  actor_user_id: string;
  role_version: number;
  reason?: string;
  before_state?: unknown;
  after_state?: unknown;
  created_at: string;
}
